### ISSUE TITLE
Weird publish output while publishing a new version? 

### ISSUE THREAD

**Issue by dotansimha** [atlassian/changesets#427]
## Affected Packages

cli?

## Problem

I'm using a custom script for setting our pre-release versions, and I noticed an error *sometimes* from NPM. The error I'm getting is:

```
error an error occurred while publishing @graphql-codegen/flow: E403 403 Forbidden - PUT https://registry.npmjs.org/@graphql-codegen%2fflow - You cannot publish over the previously published versions: 1.17.8-alpha-0f93aa04.0. 
🦋  error In most cases, you or one of your dependencies are requesting
🦋  error a package version that is forbidden by your security policy.
```

I scanned the log, and noticed that `changesets` already checks that, and it seems valid:

```
// ...
🦋  info npm info @graphql-codegen/flow
// ...
🦋  info @graphql-codegen/flow is being published because our local version (1.17.8-alpha-0f93aa04.0) has not been published on npm
// ...
error an error occurred while publishing @graphql-codegen/flow: E403 403 Forbidden - PUT https://registry.npmjs.org/@graphql-codegen%2fflow - You cannot publish over the previously published versions: 1.17.8-alpha-0f93aa04.0. 
🦋  error In most cases, you or one of your dependencies are requesting
🦋  error a package version that is forbidden by your security policy.
// ...
```

Maybe I'm missing something - but it seems like `changesets` looks for that package in that specific version, and it can't find it in NPM. But then when it tries to publish it, it fails because it already exists. 

I checked and the package does exists on NPM, around the same time the pipeline was executed. I can only assume that the publish went fine, and the NPM error reflects something else? 

## Proposed solution

No idea. Back then we had similar issues with E403 from NPM, and the messages was never informative enough, or mistaken. @ardatan suggested once that publishing many packages in short period of time causes issues, and it's better to limit the amount of `publish` command and not run all packages in parallel. We tried to limit the amount of publishing to maximum of 5 in parallel, and that seems to resolve the issue. 

@Andarist

**Comment by Andarist** [comment]
Hard to tell what might be the exact issue here - I would have to investigate the code responsible for the whole publishing process. It's important to assess if there are for example any automatic retries happening, how errors are handled etc. I can get to it, but it would probably take me some time - I'm having a lot of OSS issues in my backlog right now.

Just a note - we do limit how many internal `publish` commands are being executed in parallel, but I don't recall what's the exact limit right now. This piece of code would have to be examined as well to check for any potential logical issues.

**Comment by dotansimha** [comment]
@Andarist I'm not sure it's limited, according to this piece of code: https://github.com/atlassian/changesets/blob/master/packages/cli/src/commands/publish/publishPackages.ts#L73-L91 

I think maybe we can wrap it with https://github.com/sindresorhus/p-limit ? and add a config field for setting the limitation.
I can open a PR for that, what do you think?

**Comment by dotansimha** [comment]
Update: 
I tried to run the same thing 5 times: 4 times with the default limitation (40) and one time with modified version of it (5, using `patch-package`). All flows with 40 concurrent npm requests failed on publish command. The packages were published and seems valid, but it returns E403. With the 5 requests limit, it seems to pass correctly with no errors. 
You can see it here: https://github.com/dotansimha/graphql-code-generator/pull/4498 

I guess this issue only effect large codebases with more than a few packages, so that's why it happens for us.

**Comment by Andarist** [comment]
> Maybe we should have separate limitation for publish command, and allow to customize it with the config file?

I would start with just changing the limit - I don't see much value of having this configurable. Either a limit is good or either it is not - or maybe to rephrase: I don't see a value in being able to increase it.

I'm just wondering why this problem hasn't been surfaced earlier 🤔 There are already quite some Changesets users and I use it in a couple of multi-package repositories as well and I haven't bumped into it as well. I don't have repositories with so many packages managed by Changesets right now though, so maybe it's just that.

It hasn't been yet too problematic (or at least no one has reported issues with it), but we should handle [this issue](https://github.com/atlassian/changesets/issues/238) in the future. So assuming this is down the road - maybe we should just limit publishing to 1 at a time? It would technically be possible to implement parallel publishing in a topological order but it would increase the complexity quite a bit and I don't particularly see it being worth it. I also think that `pnpm` runs a single publish at a time given it [awaits a single publish](https://github.com/pnpm/pnpm/blob/2a41ce95c42087213ef1d54b03d68277515d2ceb/packages/plugin-commands-publishing/src/recursivePublish.ts#L89) before moving onto the next one. Could you confirm this @zkochan ? @mitchellhamilton WDYT? do we need to run publish in parallel? benefits of this seem rather negligible for me

**Comment by mitchellhamilton** [comment]
Given 5 or some limit works fine, we should probably go with that. I'm not overly concerned about the performance of publish given it's rarely going to be the bottleneck in a release workflow(especially because we encourage publishing on CI where installing dependencies is much more likely to be the bottleneck than publishing).

**Comment by Andarist** [comment]
Right, that's my exact thinking - I'm just thinking that maybe even limiting it to 1 would be OK as that would make it easier to implement publishing in a topological order as that is something that we should do as well.

**Comment by dotansimha** [comment]
It seems like 5 works fine, tested with 6 and 8, both worked fine. 10 failed. I guess 5 is ok. But if you want to go with 1, I'm fine with it either (it will have some effect on our canary release pipeline performance, since we have ~40 packages).

Btw, why topological order is needed? just for verification or something else?

**Comment by Andarist** [comment]
If A depends on B and we publish A first but fail (for whatever reason) to publish B then we might end up with a broken package A being on npm - it would potentially specify non-existent (on npm) version of B as its dep.

An alternative approach to a topological order is to publish packages in whatever order under a separate tag and switch the tag on all published packages after all packages are published. When I think about it now - this might be a better approach as it works better with cyclic dependencies 🤔

**Comment by Noviny** [comment]
I'd +1 making a PR to lower this number for publishing
